Userland/Libraries/LibJS/Runtime/VM.cpp

```diff
@@ -129,24 +129,151 @@ Symbol* VM::get_global_symbol(const String& description)
+            possible_match = scope->get_from_scope(name);
             if (possible_match.has_value()) {
-                if (!first_assignment && possible_match.value().declaration_kind == DeclarationKind::Const) {
-                    throw_exception<TypeError>(global_object, ErrorType::InvalidAssignToConst);
-                    return;
-                }
+                specific_scope = scope;
+                break;
+            }
+        }
+    }
+
+    if (specific_scope && possible_match.has_value()) {
+        if (!first_assignment && possible_match.value().declaration_kind == DeclarationKind::Const) {
+            throw_exception<TypeError>(global_object, ErrorType::InvalidAssignToConst);
+            return;
+        }
+
+        specific_scope->put_to_scope(name, { value, possible_match.value().declaration_kind });
+        return;
+    }
+
+    if (specific_scope) {
+        specific_scope->put_to_scope(name, { value, DeclarationKind::Var });
+        return;
+    }
 
-                scope->put_to_scope(name, { value, possible_match.value().declaration_kind });
+    global_object.put(name, value);
+}
+
+void VM::assign(const FlyString& target, Value value, GlobalObject& global_object, bool first_assignment, ScopeObject* specific_scope)
+{
+    set_variable(target, move(value), global_object, first_assignment, specific_scope);
+}
+
+void VM::assign(const Variant<NonnullRefPtr<Identifier>, NonnullRefPtr<BindingPattern>>& target, Value value, GlobalObject& global_object, bool first_assignment, ScopeObject* specific_scope)
+{
+    if (auto id_ptr = target.get_pointer<NonnullRefPtr<Identifier>>())
+        return assign((*id_ptr)->string(), move(value), global_object, first_assignment, specific_scope);
+
+    assign(target.get<NonnullRefPtr<BindingPattern>>(), move(value), global_object, first_assignment, specific_scope);
+}
+
+void VM::assign(const NonnullRefPtr<BindingPattern>& target, Value value, GlobalObject& global_object, bool first_assignment, ScopeObject* specific_scope)
+{
+    auto& binding = *target;
+    auto this_value = this->this_value(global_object);
+    auto this_object = this_value.to_object(global_object);
+    if (exception())
+        return;
+
+    switch (binding.kind) {
+    case BindingPattern::Kind::Array: {
+        // FIXME: "iterable" for value.
+        if (!value.is_array()) {
+            throw_exception<TypeError>(global_object, ErrorType::NotIterable, value.to_string_without_side_effects());
+            return;
+        }
+        auto& array = value.as_array();
+        auto& properties = array.indexed_properties();
+        size_t index = 0;
+        for (; index < properties.array_like_size(); ++index) {
+            if (exception())
                 return;
+
+            if (index >= binding.properties.size())
+                break;
+
+            JS::Value value { js_undefined() };
+            auto maybe_value = properties.get(this_object, index);
```

Comment:
Missing exception check - you can also use this shortcut:

```suggestion
            auto value = properties.get(this_object, index).value_or(js_undefined());
            if (exception())
                return;
```